File: src/main/java/patterns/subsets/Permutations.java
package patterns.subsets;

import java.util.*;

  /**
        Time complexity
        We know that there are a total of N! permutations of a set with ‘N’ numbers.
        In the algorithm above, we are iterating through all of these permutations with the help of the two ‘for’ loops.
        In each iteration, we go through all the current permutations to insert a new number in them on line 17 (line 23 for C++ solution).
        To insert a number into a permutation of size ‘N’ will take O(N*N!)., which makes the overall time complexity of our algorithm O(N*N!)..

        Space complexity#
        All the additional space used by our algorithm is for the result list and the queue to store the intermediate permutations.
        If you see closely, at any time, we don’t have more than N! permutations between the result list and the queue.
        Therefore, the overall space complexity to store N!N! permutations each containing NN elements will be O(N*N!).

   **/

class Permutations {

    public static List<List<Integer>> findPermutations(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();

        List<Integer> choices = new ArrayList<>();

        for(int elem : nums ){
            choices.add(elem);
        }

        doFindPermutationsRecursively(result,new ArrayList<>(),choices);


        return result;
    }

    public static List<List<Integer>> permuteBFS(int[] nums) {

        List<List<Integer>> subsets = new ArrayList<>();

        Queue<List<Integer>> permutations = new LinkedList<>();

        permutations.add(new ArrayList<>());

        for(int number : nums ) {

            int size = permutations.size();

            for(int i = 0 ; i < size; i++ ) {
                List<Integer> oldPermutations = permutations.poll();

                for(int j = 0 ; j <= oldPermutations.size(); j++ ) {

                    List<Integer> newPermutations = new ArrayList<>(oldPermutations);
                    newPermutations.add(j, number);

                    if(newPermutations.size() == nums.length ) subsets.add(newPermutations);
                    else permutations.add(newPermutations);
                }
            }
        }


        return subsets;
    }

    public static void doFindPermutationsRecursively(List<List<Integer>> result,
                                                     List<Integer> prefix, List<Integer> choices) {

        if(choices.size() == 0 ) {
            result.add(new ArrayList<>(prefix));
        } else {
            for (int i = 0; i < choices.size(); i++) {

                prefix.add(choices.get(i));
                int elem = choices.remove(i);
                doFindPermutationsRecursively(result, prefix, choices);
                prefix.remove(prefix.size() - 1);
                choices.add(i, elem);
            }
        }

    }

    public static void main(String[] args) {
        long startTime = System.currentTimeMillis();
        List<List<Integer>> result = Permutations.findPermutations(new int[] { 0,1, 2,3,4,5,6,7,8,9});
        long endTime = System.currentTimeMillis();
        System.out.println("Elapse time recursive is  " + (endTime - startTime)/10);
        long startTime2 = System.currentTimeMillis();
        List<List<Integer>> result2 = Permutations.permuteBFS(new int[] { 0,1, 2,3,4,5,6,7,8,9 });
        long endTime2 = System.currentTimeMillis();
        System.out.println("Elapse time bfs approach is  " + (endTime2 - startTime2)/10);
//        System.out.println("Here are all the permutations: " + result);
//        System.out.print("Here are all the permutations: " + result2);
    }
}

File: src/main/scala/patterns/Permutations.scala
package patterns

import patterns.subsets.Permutations

import java.util.List
import scala.collection.mutable.Queue
import scala.collection.mutable.ArrayBuffer
import scala.collection.mutable.LinkedList

import java.util

object Permutations {

  def main(args: Array[String]): Unit = {
    val startTime = System.currentTimeMillis
    val result = Permutations.findPermutations(Array[Integer](0, 1, 2, 3, 4, 5, 6, 7, 8, 9))
    val endTime = System.currentTimeMillis
    System.out.println("Elapse time recursive is  " + (endTime - startTime))
    val startTime2 = System.currentTimeMillis
    val result2 = Permutations.permuteBFS(Array[Integer](0, 1, 2, 3, 4, 5, 6, 7, 8, 9))
    val endTime2 = System.currentTimeMillis
    System.out.println("Elapse time bfs approach is  " + (endTime2 - startTime2))
  }



 def findPermutations(nums: Array[Integer]) : ArrayBuffer[ArrayBuffer[Integer]] = {

   var result = new ArrayBuffer[ArrayBuffer[Integer]]

   var choices = new ArrayBuffer[Integer]()

   for(i <- nums.indices ) {

     choices += nums(i)
   }

   doFindPermute(result,choices,new ArrayBuffer[Integer])
   result
  }

  def doFindPermute(result : ArrayBuffer[ArrayBuffer[Integer]],
                        choices: ArrayBuffer[Integer],prefix: ArrayBuffer[Integer] ) : Unit ={


    if(choices.isEmpty ){
      val temp = new ArrayBuffer[Integer]()
      prefix.copyToBuffer(temp)
      result += temp
    } else {

      for(i <- choices.indices) {

        prefix += choices(i)
        var elem = choices.remove(i)
        doFindPermute(result,choices,prefix)
        choices.insert(i,elem)
        prefix.remove(prefix.size - 1)

      }
    }
  }


  def permuteBFS(nums: Array[Integer]) : ArrayBuffer[ArrayBuffer[Integer]] = {


    val subsets = new ArrayBuffer[ArrayBuffer[Integer]]()

    val permutations: Queue[ArrayBuffer[Integer]] = new Queue[ArrayBuffer[Integer]]()

    permutations += (new ArrayBuffer[Integer]())

    for(number <- nums){

      val size = permutations.size

      for(i <- 0 until size ) {

        val oldPermutations = permutations.dequeue()

        for(j <- 0 to oldPermutations.size ) {

          val newPermutations = new ArrayBuffer[Integer]()
          oldPermutations.copyToBuffer(newPermutations)

          newPermutations.insert(j,number)

          if(newPermutations.size == nums.length ) subsets += newPermutations
          else permutations += newPermutations
        }
      }
    }
    subsets
  }


  def doPermuteBFS(nums: Array[Integer]) : util.ArrayList[util.ArrayList[Integer]] = {


    val subsets = new util.ArrayList[util.ArrayList[Integer]]()

    val permutations: util.LinkedList[util.ArrayList[Integer]] = new util.LinkedList[util.ArrayList[Integer]]()

    permutations.add(new util.ArrayList[Integer]())

    for(number <- nums){

      val size = permutations.size

      for(i <- 0 until size ) {

        val oldPermutations = permutations.poll()

        for(j <- 0 to oldPermutations.size ) {

          val newPermutations = new util.ArrayList[Integer](oldPermutations)

          newPermutations.add(j,number)

          if(newPermutations.size == nums.length ) subsets.add( newPermutations)
          else permutations.add(newPermutations)
        }
      }
    }
    subsets
  }

}
